web_advanced_filter/models/ir_filters.py

```diff
@@ -66,70 +72,87 @@ def _compute_is_frozen(self):
+                _logger.error(
+                    "Unknown model %s used in filter %d",
+                    this.model_id,
+                    this.id,
+                )
                 continue
-            domain = eval_n(this.domain_this)
+            domain = self._eval_domain(this.domain_this)
             for u in this.union_filter_ids:
+                if u.model_id != this.model_id:
+                    _logger.warning(
+                        "Model mismatch in helper %d on filter %d!",
+                        u.model_id,
+                        this.model_id,
+                    )
+                    continue
                 if u.evaluate_always:
-                    matching_ids = self.env[u.model_id].search(
-                        eval_n(u.domain),
-                    ).ids
+                    matching_ids = (
+                        self.env[this.model_id]
+                        .search(self._eval_domain(u.domain))
+                        .ids
+                    )
                     domain = expression.OR([
                         domain,
                         [('id', 'in', matching_ids)],
                     ])
                 else:
-                    domain = expression.OR([domain, eval_n(u.domain)])
+                    domain = expression.OR(
+                        [domain, self._eval_domain(u.domain)]
+                    )
             for c in this.complement_filter_ids:
+                if c.model_id != this.model_id:
+                    _logger.warning(
+                        "Model mismatch in helper %d on filter %d!",
+                        c.model_id,
+                        this.model_id,
+                    )
+                    continue
                 if c.evaluate_before_negate:
-                    matching_ids = self.env[c.model_id].search(
-                        eval_n(c.domain),
-                    ).ids
+                    matching_ids = (
+                        self.env[this.model_id]
+                        .search(self._eval_domain(c.domain))
+                        .ids
+                    )
                     domain = expression.AND([
                         domain,
                         [('id', 'not in', matching_ids)],
                     ])
                 else:
                     domain = expression.AND([
                         domain,
-                        ['!'] + eval_n(c['domain']),
+                        ["!"] + self._eval_domain(c["domain"])
                     ])
-            this.domain = str(expression.normalize_domain(domain))
+            this.domain = repr(expression.normalize_domain(domain))
 
     @api.multi
     def _inverse_domain(self):
         for this in self:
             this.domain_this = this.domain
 
     @api.multi
-    @api.depends('domain')
+    @api.depends('domain', 'model_id')
     def _compute_flags(self):
         """check if this filter contains references to x2many fields. If so,
```

Comment:
This is an explanation that really does not explain anything. What goes wrong with negation? What does negation mean here? How is this solved by removing resulting ids?

Reply:
I'm having trouble figuring out the history of this.

It was added in #83, with a reference to [this `therp-backports` bug report](https://bugs.launchpad.net/therp-backports/+bug/1062795). But I can't reproduce the original issue, neither in OpenERP 7 nor in Odoo 12. I'm trying it by directly running running variants of `res.partner.search([("category_id", "!=", ...)])`. It's possible it's no longer necessary, but maybe I'm just missing the reason it was necessary originally.

#100 dealt with `auto_join`. I can't reproduce the underlying issue for that either, though maybe I'm interpreting it wrong—it sounds to me like `res.partner.search(['|', ('name', '=', 'something'), ('message_ids', '=', some_id)])` would only return results that match the second term, because `message_ids` has `auto_join`, but it also returns records that only match the first term, like you'd expect.

It almost seems like this stuff wasn't necessary for version 7, in which case it wouldn't be necessary for 12 either.

Do you have any insight?

Reply:
I have no insights to add here. Maybe just add tests for the problems that were supposedly solved, and see that they do not re-occur. It would be good if the code can be drastically simplified, or organized in such a way that it is more understandable.

Reply:
@janverb to reproduce the first one, you need a partner with two labels set, and say `[("category_id.name", "!=", "label1")]`. You won't get the inverse of `[("category_id.name", "=", "label1")]`. So the issue is how Odoo treats dotted paths in domains.

The second is similar, when you have a path of auto_joined fields and union this with a domain that doesn't, you expect the resulting records of both filters. But if you put this in one domain, the resulting statement is a join that only yields results of the join. You should find an issue by the customer a from around the time of the original commit with an example.

Reply:
@hbrunn Thank you! I've been able to reproduce both issues on version 12. I'll write tests and do some cleanup.

I'm still unsure about `has_x2many |= bool(field.compute)` and `has_auto_join |= bool(field.compute)`. Are those there to be more careful about fields with custom searching logic? If so, should they check for `field.search` instead?

Reply:
yes, that's for custom searching and yes, you'll be better off checking for explicit search functions

Reply:
Something else I'm now wondering about: shouldn't the `domain_this` of the main filter be subject to the same check for autojoined fields? It doesn't matter whether the autojoined search is in a union helper filter, or in the main filter, or even in a complement filter, as long as it's getting ORed at some point.

It might be a safer, simpler setup to merge `evaluate_before_negate` and `evaluate_always` into a single field and look at it to evaluate any of the own domain and the domains of the union and complement filters.

Reply:
A redesign seems out of scope for the migration so I've just cleaned it up and added more documentation and tests.